kernel/aster-nix/src/syscall/madvise.rs

```diff
@@ -33,8 +33,17 @@ pub fn sys_madvise(
 }
 
 fn madv_free(start: Vaddr, len: usize, ctx: &Context) -> Result<()> {
-    debug_assert!(start % PAGE_SIZE == 0);
-    debug_assert!(len % PAGE_SIZE == 0);
+    if start % PAGE_SIZE != 0 {
+        return_errno_with_message!(Errno::EINVAL, "the start address should be page aligned");
+    }
+
+    if len % PAGE_SIZE != 0 {
+        return_errno_with_message!(Errno::EINVAL, "the length should be page aligned");
+    }
+
+    if len == 0 {
```

Comment:
Will Linux report errors when these conditions are not met? Are these behaviors documented by the man page or confirmed on Linux? I mean, for example, it is possible that Linux decides that if `len == 0`, the the system call should do nothing, instead of reporting an error.

Reply:
According to the [Linux man page of madvise](https://linux.die.net/man/2/madvise), the `EINVAL` is set when 
- addr is not page-aligned.
- The value len is negative. (?)

And in the [source code of Linux](https://elixir.bootlin.com/linux/v6.10.5/source/mm/madvise.c#L1417), it checks the `addr` and the `len` with code

```
if (!PAGE_ALIGNED(start))
	return -EINVAL;
len = PAGE_ALIGN(len_in);

/* Check to see whether len was rounded up from small -ve to zero */
if (len_in && !len)
	return -EINVAL;
end = start + len;
if (end < start)
	return -EINVAL;

if (end == start)
	return 0;
```

It seems the `end == start`  is the `len == 0` condition, and this system call does nothing.

However, the these checks is in the `sys_madvise` in linux, shall I change the check code location?

Reply:
If these sanity checks are applicable to all sub-commands of madvice, then yes, they should be moved to the `sys_madvise` function. And let's try to follow Linux behaviors as closely as possible. For example, no error is reported if the length is not page-aligned.